File: xmlBlaster/demo/perl/xmlrpc/hello.pl
#!/usr/bin/perl
# Name: xmlBlaster/demo/perl/xmlrpc/hello.pl (xmlBlaster.org)
# Invoke
#   perl hello.pl http://myHost:8080/
# if xmlBlaster runs on 'myHost'

use Frontier::Client;
use MIME::Base64;
#use String::Trim;
    
#$server_url = 'http://MyHost:8080/';
$server_url = @ARGV[0];
if ($#ARGV == -1) {
   $host = `uname -n`;
   $host =~ s/^\s*(.*?)\s*$/$1/; # trim whitespace
   $server_url = "http://" . $host . ":8080/";  # guess where xmlBlaster is
}
print "\nTrying to connect to xmlBlaster server on $server_url ...\n";

# Make an object to represent the XMLRPC server.
$server = Frontier::Client->new(url => $server_url);
print "Connected to xmlBlaster server on $server_url \n";

# Call the remote server and get our result.
$sessionId = $server->call('authenticate.login', "ben", "secret",
                                                 "<qos></qos>", "");
print "Login success, got secret sessionId=$sessionId \n";

# Call the server and get its current memory consumption.
$queryKey = "<key oid='__cmd:?totalMem'/>";
# Call the server and query all messages with XPath:
#$queryKey = "<key queryType='XPATH'>/xmlBlaster</key>";

@msgUnits = $server->call('xmlBlaster.get', $sessionId, $queryKey, "<qos/>");
print "\nResults for a get($queryKey):";
for $i (0 .. $#msgUnits) {
   for $j (0 .. $#{$msgUnits[$i]}) {
      print "\n-------------#$j-------------------";
      $key = $msgUnits[$i][$j][0];
      $contentBase64AndEncoded = $msgUnits[$i][$j][1];
      $content = decode_base64($contentBase64AndEncoded->value());
      $qos = $msgUnits[$i][$j][2];
      print $key;
      print "\n<content>" . $content . "</content>\n";
      print $qos;
      print "\n-------------#$j-------------------\n";
   }
}

# Try publishing a message:
$returnQos = $server->call('xmlBlaster.publish', $sessionId,
                           "<key oid='MyMessage'></key>",
                           "Hello world", "<qos><forceUpdate /></qos>");
print "\nResult for a publish():\n------------", $returnQos, "\n------------\n";


# Logout from xmlBlaster
$server->call('authenticate.logout', $sessionId);
print "\nLogout done, bye.\n";

File: xmlBlaster/demo/python/hello.py
# -----------------------------------------------------
# Python hello world demo xmlrpc-client
# Invoke:  python  hello.py  http://myhost:8080
# -----------------------------------------------------

import xmlrpclib
import sys

# The url to xmlBlaster (see the output of xmlBlaster server):
server_url = sys.argv[1]


# -----------------------------------------------------
print "\nTrying to connect to xmlBlaster server on ", server_url, " ...\n"
# Create an object to represent our server.
server = xmlrpclib.Server(server_url)


# -----------------------------------------------------
# Login to xmlBlaster using a password file:
#qos = "<qos><securityService type='htpasswd' version='1.0'> \
#           <![CDATA[ \
#             <user>michele</user> \
#             <passwd>secret</passwd> \
#           ]]> \
#        </securityService></qos>";

# Login to xmlBlaster using default authentication:
qos = "<qos></qos>"

sessionId = server.authenticate.login( "ben", "secret", qos, "")
print "\nLogin success with on ", server_url, " sessionId=", sessionId


# -----------------------------------------------------
# Call the server and get its current memory consumption.
print "\n*  Trying to access the consumed memory of the xmlBlaster server ..."
messages = server.xmlBlaster.get( sessionId, "<key oid='__cmd:?totalMem'></key>", "<qos></qos>")
print "   Received ", len(messages), " messages:"
for msg in messages:
   key = msg[0]
   content = msg[1]     # content is of type xmlrpclib.Binary
   qos = msg[2]
   print "      key=", key
   print "      content=", content.data, " bytes"
   print "      qos=", qos


# -----------------------------------------------------
print "\n*  Trying to publish a 'Hello world' message ..."
# How do i generate a binary content with python??
#content = xmlrpclib.Binary(R"Hello world")
#server.xmlBlaster.publish( sessionId, "<key oid='MyMessage'></key>", content.data, "<qos></qos>")

content = "Hello world"
server.xmlBlaster.publish( sessionId, "<key oid='MyMessage'></key>", content, "<qos></qos>")
print "   Publish success for message 'MyMessage' content=", content


# -----------------------------------------------------
# Call the server and get the hello world message again
print "\n*  Trying to access our previous published 'Hello world' message ..."
messages = server.xmlBlaster.get( sessionId, "<key oid='MyMessage'></key>", "")
print "   Received ", len(messages), " messages:"
for msg in messages:
   key = msg[0]
   content = msg[1]     # content is of type xmlrpclib.Binary
   qos = msg[2]
   print "      key=", key
   print "      content=", content.data
   print "      qos=", qos


# -----------------------------------------------------
# Asynchronous access - not yet implemented
# We need to instantiate a Python xmlRpc callback server first
#server.xmlBlaster.subscribe( sessionId, "<key oid='__cmd:?totalMem'></key>", "<qos></qos>")


# -----------------------------------------------------
# Logout from xmlBlaster
server.authenticate.logout( sessionId ) #server.authenticate.logout( sessionId, "<qos></qos>" )
print "\nLogout done, bye.\n"
